LLK/scratch-vm
# Various fixes to extension disconnect errors

### Resolves

- Resolves #1590: Consider adding a way to detect that wedo2 has disconnected?
- Fixes disconnect error handling in existing hardware extensions to correctly clear timeouts and other data

### Notes

- In implementing this fix for LEGO WeDo2, I noticed that error handling wasn't quite complete among all the hardware extensions.  This PR changes BLE/BT sockets to take in `disconnectCallback`s in order to properly have hardware extensions clear any remaining timeoutIDs on the window, clear locally stored data, etc.  As a result of those enhancements, this PR requires testing all three extensions.
- Disconnect alerts might look too large, that's being addressed in another PR
- Once this lands, then https://github.com/LLK/scratch-gui/pull/3922 can land as well

### Testing LEGO WeDo 2
1. Load extension and connect LEGO WeDo2
2. Turn off LEGO WeDo2 by pressing button for 3 seconds
3. In 5 seconds or less, should see disconnect alert and status button should turn orange

### Testing EV3
1. Load extension and connect EV3
2. Turn off EV3
3. In a few seconds, should see disconnect alert and status button should turn orange

### Testing micro:bit
1. Load extension and connect micro:bit
2. Turn off micro:bit
3. In 5 seconds, should see disconnect alert and status button should turn orange
